Add user offline tuning doc (#1305)
ROCm/hipBLASLt: docs/how-to-use-hipblaslt-offline-tuning.rst

.. meta::
   :description: A library that provides GEMM operations with flexible APIs and extends functionalities beyond the traditional BLAS library

>    :keywords: hipBLASLt, ROCm, library, API, tool

Reviewer: ```suggestion
   :keywords: hipBLASLt, ROCm, library, API, tool, tuning, GEMM
```

.. _how-to-use-hipblaslt-offline-tuning:

********************************
User Offline Tuning
********************************

``hipblaslt-bench`` can be used to find the best performing GEMM kernel for each of a given set of GEMM problems. 

> It has a command line interface. (See :ref:`clients` for more details.)

Reviewer: ```suggestion
``hipblaslt-bench`` can be used to find the best-performing GEMM kernel for a given set of GEMM problems. Use the command line interface to access this functionality. (See :ref:`clients` for more details.)
```

> Through the User Offline Tuning mechanism, the best solution index for a given problem size obtained by ``hipblaslt-bench`` can be directly used in future GEMM calls.

Reviewer: Should "User Offline Tuning" be fully capitalized?

Author: Yes

> However, please note that these indices cannot be reused across library releases or across different device architectures.

Reviewer: ```suggestion
``hipblaslt-bench`` provides the best solution index for a given problem size. This index can be used directly in future GEMM calls through the User Offline Tuning mechanism. However, these indices cannot be reused across library releases or across different device architectures.
```

> Running

Reviewer: ```suggestion
Using hipblaslt-bench to run the tuning with the best GEMM kernel
```

Reviewer: I was looking for something more descriptive than just "running". Is this a good overview of the section?

=================================

> Follow these steps:

Reviewer: ```suggestion
To find and use the best GEMM kernel for a problem, follow these steps:
```

> 1. Generate tuning command line by setting environment variable ``HIPBLASLT_LOG_MASK=32`` before calling hipBLASLt APIs if you are unsure how to use ``hipblaslt-bench``. (see :ref:`Logging and heuristics <logging-heuristics>` for more details.)

Reviewer: ```suggestion
1. Generate the tuning command line by setting the environment variable ``HIPBLASLT_LOG_MASK=32`` before calling any hipBLASLt APIs. For more details on how to use ``hipblaslt-bench``, see :ref:`Logging and heuristics <logging-heuristics>`.
```

> 

Reviewer: In this section, indent the steps so they align with the numbering. The dots should align with "Generate" not with the left margin. I tried to add the indent to my changes below, but you'll need to adjust the other lines.

> - In Bash shell:

Reviewer: ```suggestion
   In the Bash shell, set the following environment variable:
```

.. code-block:: bash

    export HIPBLASLT_LOG_MASK=32

> - Using `sample_hipblaslt_gemm.cpp <https://github.com/ROCm/hipBLASLt/blob/develop/clients/samples/01_basic_gemm/sample_hipblaslt_gemm.cpp>`_  as an example:

Reviewer: ```suggestion
    In this case, `sample_hipblaslt_gemm.cpp <https://github.com/ROCm/hipBLASLt/blob/develop/clients/samples/01_basic_gemm/sample_hipblaslt_gemm.cpp>`_ is used as an example:
```

.. code-block:: bash

    ./sample_hipblaslt_gemm

> - The tuning command will log:

Reviewer: ```suggestion
   The tuning command displays the following log entry:
```

.. code-block:: bash

    hipblaslt-bench --api_method c -m 1024 -n 512 -k 1024 --lda 1024 --ldb 1024 --ldc 1024 --ldd 1024  --stride_a 0 --stride_b 0 --stride_c 0 --stride_d 0  --alpha 1.000000 --beta 1.000000 --transA N --transB N --batch_count 1  --a_type f16_r --b_type f16_r --c_type f16_r --d_type f16_r --scale_type f32_r --bias_type f32_r   --compute_type f32_r 

> 2. Set the environment variable ``HIPBLASLT_TUNING_FILE=<file_name>`` to tune and store the tuning result of the best solution indices for those GEMM problems, where ``<file_name>`` points to the tuning file.

Reviewer: ```suggestion
2. Set the environment variable ``HIPBLASLT_TUNING_FILE=<file_name>`` to tune and store the tuning result of the best solution indices for the GEMM problems. The ``<file_name>`` points to the tuning file.
```

> 2. Set the environment variable ``HIPBLASLT_TUNING_FILE=<file_name>`` to tune and store the tuning result of the best solution indices for those GEMM problems, where ``<file_name>`` points to the tuning file.

Reviewer: We should probably include an example of how to use this as a bash command (the same way we demonstrated how to set the environment variable above).

> - Please note that the default settings will have the following changes in the tuning environment.

Reviewer: ```suggestion
   The following parameters can be used to override the default settings in the tuning environment.
```

> .. code-block:: bash

Reviewer: This is a little confusing. Can the user append one or more of these parameters to the environment variable. Probably better to explain. The output doesn't align very well. I think you can eliminate the second tab here (before the default values) for display purposes.

If this is taken directly from the help, we can otherwise leave it as is.

Author: Yes, this is taken directly from the help.
This content is intended to express that the default settings for the following parameters in hipblaslt-bench will be changed in the tuning environment.

    --iters |-i <value>             Iterations to run inside timing loop                                                (Default value is: 1000)
    --cold_iters |-j <value>        Cold Iterations to run before entering the timing loop                              (Default value is: 1000)
    --algo_method <value>           Use different algorithm search API. Options: heuristic, all, index (still considered all).                  (Default value is: all)
    --requested_solution <value>    Requested solution num. Set to -1 to get all solutions. Only valid when algo_method is set to heuristic.    (Default value is: -1)
    --rotating <value>              Use rotating memory blocks for each iteration, size in MB.                          (Default value is: 512)

> - After the tuning completes, the expected output is as follows:

Reviewer: ```suggestion
   After the tuning completes, the expected output is displayed as follows:
```

.. code-block:: bash

>     ./hipblaslt-bench --api_method c -m 64 -n 64 -k 1024 --lda 64 --ldb 1024 --ldc 64 --ldd 64  --stride_a 0 --stride_b 0 --stride_c 0 --stride_d 0  --alpha 1.100000 --beta 0.900000 --transA N --transB N --batch_count 1  --a_type f32_r --b_type f32_r --c_type f32_r --d_type f32_r --scale_type f32_r --bias_type f32_r   --compute_type f32_r

Reviewer 2: with yangwen's PR, the actual output of MASK=32 would contain the solution_index.

Author: it’s done in commit [0091f2a](https://github.com/ROCm/hipBLASLt/pull/1305/commits/0091f2a502cb324e7145fda9ec084ad580517ddb)

    
    Winner: 
    transA,transB,grouped_gemm,batch_count,m,n,k,alpha,lda,stride_a,beta,ldb,stride_b,ldc,stride_c,ldd,stride_d,a_type,b_type,c_type,d_type,compute_type,scaleA,scaleB,scaleC,scaleD,amaxD,activation_type,bias_vector,bias_type,rotating_buffer,hipblaslt-Gflops,hipblaslt-GB/s,us,soulution_index
    N,N,0,1,64,64,1024,1.1,64,65536,0.9,1024,65536,64,4096,64,4096,f32_r,f32_r,f32_r,f32_r,f32_r,0,0,0,0,0,none,0,f32_r,512,903.229,55.8608,9.28735,49257

> 3. Set the environment variable ``HIPBLASLT_TUNING_OVERRIDE_FILE=<file_name>`` to load the tuning file and override the default kernel selection with the kernels found, where ``<file_name>`` points to the tuning file.

Reviewer: ```suggestion
3. Set the environment variable ``HIPBLASLT_TUNING_OVERRIDE_FILE=<file_name>`` to load the tuning file and override the default kernel selection with the optimal kernel choices, where ``<file_name>`` points to the tuning file.
```

> 3. Set the environment variable ``HIPBLASLT_TUNING_OVERRIDE_FILE=<file_name>`` to load the tuning file and override the default kernel selection with the kernels found, where ``<file_name>`` points to the tuning file.

Reviewer: Again, we probably want to highlight the command to run in bash syntax on a separate line, as in section 1

> - For example, user can use ``hisblaslt-bench`` (algo_method: heuristic) to obtain solutions for a problem, which will include the best tuning solution index.

Reviewer: ```suggestion
   For example, a user can use ``hisblaslt-bench`` (with ``algo_method`` set to ``heuristic``) to obtain the solutions for a problem, which include the best tuning solution index.
```

Reviewer: Are the solutions below (0, 1, and 2) ranked in order, where 0 is the best solution? Also, is it possible for the user to save the hipblaslt-bench command displayed below and run it again if they want?

Author: Yes, the tuning result of the best solution will be placed at position zero.
